packages/macro/index.test-d.tsx

```diff
@@ -0,0 +1,334 @@
+  })
+)
+
+// @ts-expect-error id or message should be presented
+expectType<MessageDescriptor>(defineMessage({}))
+
+///////////////////
+//// Plural  //////
+///////////////////
+
+expectType<string>(
+  plural("5", {
+    // @ts-expect-error extra properties are not allowed
+    incorrect: "",
+
+    one: "...",
+    other: "...",
+    few: "...",
+    many: "...",
+    zero: "...",
+  })
+)
+
+expectType<string>(
+  plural(5, {
+    one: "...",
+    other: "...",
+  })
+)
+
+// with offset
+expectType<string>(
+  plural(5, {
+    one: "...",
+    other: "...",
+    offset: 5,
+  })
+)
+
+// exact choices
+expectType<string>(
+  plural(5, {
+    0: "...",
+    1: "...",
+    one: "...",
+    other: "...",
+  })
+)
+
+expectType<string>(
+  plural(5, {
+    // @ts-expect-error: should accept only strings
+    one: 5,
+    // @ts-expect-error: should accept only strings
+    other: 5,
+  })
+)
+
+///////////////////
+//// Select Ordinal
+///////////////////
+
+expectType<string>(
+  selectOrdinal("5", {
+    // @ts-expect-error extra properties are not allowed
+    incorrect: "",
+
+    one: "...",
+    other: "...",
+    few: "...",
+    many: "...",
+    zero: "...",
+  })
+)
+
+expectType<string>(
+  selectOrdinal(5, {
+    one: "...",
+    other: "...",
+  })
+)
+
+// with offset
+expectType<string>(
+  selectOrdinal("5", {
+    one: "...",
+    other: "...",
+    offset: 5,
+  })
+)
+
+// exact choices
+expectType<string>(
+  selectOrdinal(5, {
+    0: "...",
+    1: "...",
+    one: "...",
+    other: "...",
+  })
+)
+
+expectType<string>(
+  selectOrdinal(5, {
+    // @ts-expect-error: should accept only strings
+    one: 5,
+    // @ts-expect-error: should accept only strings
+    other: 5,
+  })
+)
+
+///////////////////
+//// Select
+///////////////////
+
+const gender = "male"
+expectType<string>(
+  select(gender, {
+    // todo: here is inconsistency between jsx macro and js.
+    //   in JSX macro you should prefix exact choices with "_"
+    //   but here is not. And it's better to use it with underscore,
+    //   because this could be statically checked by TS
+    //   type UnderscoreValue = `_${string}`;
+    male: "he",
+    female: "she",
+    other: "they",
+  })
+)
+
+expectType<string>(
+  // @ts-expect-error value could be strings only
+  select(5, {
+    male: "he",
+    female: "she",
+    other: "they",
+  })
+)
+
+expectType<string>(
+  select("male", {
+    // @ts-expect-error: should accept only strings
+    male: 5,
+    // @ts-expect-error: should accept only strings
+    other: 5,
+  })
+)
+
+// @ts-expect-error: is never read
+let m: any
+
+///////////////////
+//// JSX Trans
+///////////////////
+
+m = <Trans>Message</Trans>
+m = (
+  <Trans id="custom.id" comment="comment" context="context">
+    Message
+  </Trans>
+)
+
+// @ts-expect-error: children is required here
+m = <Trans id="custom.id" comment="comment" context="context" />
+
+// todo: write tests for
+//  render prop
+//  i18n
```

Comment:
Are you going to add these tests?

Reply:
https://github.com/lingui/js-lingui/issues/1433 i created an issue for that (task number 2)

Reply:
removed todo